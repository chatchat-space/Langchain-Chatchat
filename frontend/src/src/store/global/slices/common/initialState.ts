import { AppRouterInstance } from 'next/dist/shared/lib/app-router-context.shared-runtime';

export enum SidebarTabKey {
  Chat = 'chat',
  Market = 'market',
  Setting = 'settings',
}

export enum SettingsTabs {
  About = 'about',
  Agent = 'agent',
  Common = 'common',
  LLM = 'llm',
  TTS = 'tts',
}

export interface Guide {
  // Topic 引导
  topic?: boolean;
}

export interface GlobalCommonState {
  hasNewVersion?: boolean;
  isMobile?: boolean;
  latestVersion?: string;
  router?: AppRouterInstance;
  sidebarKey: SidebarTabKey;
}

export const initialCommonState: GlobalCommonState = {
  isMobile: false,
  sidebarKey: SidebarTabKey.Chat,
};
