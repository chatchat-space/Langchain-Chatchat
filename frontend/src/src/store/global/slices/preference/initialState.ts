import { SessionDefaultGroup, SessionGroupId } from '@/types/session';

export interface Guide {
  // Topic 引导
  topic?: boolean;
}

export interface GlobalPreference {
  // which sessionGroup should expand
  expandSessionGroupKeys: SessionGroupId[];
  guide?: Guide;
  inputHeight: number;
  mobileShowTopic?: boolean;
  sessionsWidth: number;

  showChatSideBar?: boolean;
  showSessionPanel?: boolean;
  showSystemRole?: boolean;
  telemetry: boolean | null;
  /**
   * whether to use cmd + enter to send message
   */
  useCmdEnterToSend?: boolean;
}

export interface GlobalPreferenceState {
  /**
   *  用户偏好的 UI 状态
   *  @localStorage
   */
  preference: GlobalPreference;
}

export const initialPreferenceState: GlobalPreferenceState = {
  preference: {
    expandSessionGroupKeys: [SessionDefaultGroup.Pinned, SessionDefaultGroup.Default],
    guide: {},
    inputHeight: 200,
    mobileShowTopic: false,
    sessionsWidth: 320,
    showChatSideBar: true,
    showSessionPanel: true,
    showSystemRole: false,
    telemetry: null,
    useCmdEnterToSend: false,
  },
};
