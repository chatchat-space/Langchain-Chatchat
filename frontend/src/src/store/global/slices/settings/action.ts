import { ThemeMode } from 'antd-style';
import isEqual from 'fast-deep-equal';
import { DeepPartial } from 'utility-types';
import type { StateCreator } from 'zustand/vanilla';

import { userService } from '@/services/user';
import type { GlobalStore } from '@/store/global';
import { LobeAgentSettings } from '@/types/session';
import { GlobalLLMConfig, GlobalLLMProviderKey, GlobalSettings } from '@/types/settings';
import { difference } from '@/utils/difference';
import { merge } from '@/utils/merge';

/**
 * 设置操作
 */
export interface SettingsAction {
  importAppSettings: (settings: GlobalSettings) => Promise<void>;
  resetSettings: () => Promise<void>;
  setModelProviderConfig: <T extends GlobalLLMProviderKey>(
    provider: T,
    config: Partial<GlobalLLMConfig[T]>,
  ) => Promise<void>;
  setSettings: (settings: DeepPartial<GlobalSettings>) => Promise<void>;
  switchThemeMode: (themeMode: ThemeMode) => Promise<void>;
  toggleProviderEnabled: (provider: GlobalLLMProviderKey, enabled: boolean) => Promise<void>;
  updateDefaultAgent: (agent: DeepPartial<LobeAgentSettings>) => Promise<void>;
}

export const createSettingsSlice: StateCreator<
  GlobalStore,
  [['zustand/devtools', never]],
  [],
  SettingsAction
> = (set, get) => ({
  importAppSettings: async (importAppSettings) => {
    const { setSettings } = get();
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { password: _, ...settings } = importAppSettings;

    await setSettings(settings);
  },
  resetSettings: async () => {
    await userService.resetUserSettings();
    await get().refreshUserConfig();
  },
  setModelProviderConfig: async (provider, config) => {
    await get().setSettings({ languageModel: { [provider]: config } });
  },
  setSettings: async (settings) => {
    const { settings: prevSetting, defaultSettings } = get();

    const nextSettings = merge(prevSetting, settings);

    if (isEqual(prevSetting, nextSettings)) return;

    const diffs = difference(nextSettings, defaultSettings);

    await userService.updateUserSettings(diffs);
    await get().refreshUserConfig();
  },

  switchThemeMode: async (themeMode) => {
    await get().setSettings({ themeMode });
  },
  toggleProviderEnabled: async (provider, enabled) => {
    await get().setSettings({ languageModel: { [provider]: { enabled } } });
  },
  updateDefaultAgent: async (defaultAgent) => {
    await get().setSettings({ defaultAgent });
  },
});
