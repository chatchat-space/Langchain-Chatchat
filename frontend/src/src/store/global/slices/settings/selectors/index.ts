export { modelProviderSelectors } from './modelProvider';
export { settingsSelectors } from './settings';
