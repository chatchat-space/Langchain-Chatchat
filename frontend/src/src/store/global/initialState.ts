import { GlobalCommonState, initialCommonState } from './slices/common/initialState';
import { GlobalPreferenceState, initialPreferenceState } from './slices/preference/initialState';
import { GlobalSettingsState, initialSettingsState } from './slices/settings/initialState';

export { SettingsTabs, SidebarTabKey } from './slices/common/initialState';

export type GlobalState = GlobalCommonState & GlobalSettingsState & GlobalPreferenceState;

export const initialState: GlobalState = {
  ...initialCommonState,
  ...initialSettingsState,
  ...initialPreferenceState,
};
