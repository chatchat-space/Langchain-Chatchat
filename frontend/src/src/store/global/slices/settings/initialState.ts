import { DeepPartial } from 'utility-types';

import { DEFAULT_SETTINGS } from '@/const/settings';
import { GlobalServerConfig, GlobalSettings } from '@/types/settings';

export interface GlobalSettingsState {
  avatar?: string;
  defaultSettings: GlobalSettings;
  serverConfig: GlobalServerConfig;
  settings: DeepPartial<GlobalSettings>;
  userId?: string;
}

export const initialSettingsState: GlobalSettingsState = {
  defaultSettings: DEFAULT_SETTINGS,
  serverConfig: {
    telemetry: {},
  },
  settings: {},
};
