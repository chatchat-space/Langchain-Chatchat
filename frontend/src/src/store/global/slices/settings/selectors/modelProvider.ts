import { produce } from 'immer';

import {
  AnthropicProvider,
  BedrockProvider,
  GoogleProvider,
  LOBE_DEFAULT_MODEL_LIST,
  MistralProvider,
  MoonshotProvider,
  OllamaProvider,
  OpenAIProvider,
  PerplexityProvider,
  ZhiPuProvider,
} from '@/config/modelProviders';
import { ChatModelCard, ModelProviderCard } from '@/types/llm';
import { GlobalLLMProviderKey } from '@/types/settings';
import { parseModelString } from '@/utils/parseModels';

import { GlobalStore } from '../../../store';
import { currentSettings } from './settings';

const modelProvider = (s: GlobalStore) => currentSettings(s).languageModel;
const providerEnabled = (provider: GlobalLLMProviderKey) => (s: GlobalStore) =>
  currentSettings(s).languageModel[provider]?.enabled || false;

const openAIConfig = (s: GlobalStore) => modelProvider(s).openAI;

const openAIAPIKey = (s: GlobalStore) => openAIConfig(s).OPENAI_API_KEY;
const openAIProxyUrl = (s: GlobalStore) => openAIConfig(s).endpoint;

const enableZhipu = (s: GlobalStore) => modelProvider(s).zhipu.enabled;
const zhipuAPIKey = (s: GlobalStore) => modelProvider(s).zhipu.apiKey;
const zhipuProxyUrl = (s: GlobalStore) => modelProvider(s).zhipu.endpoint;

const enableBedrock = (s: GlobalStore) => modelProvider(s).bedrock.enabled;
const bedrockConfig = (s: GlobalStore) => modelProvider(s).bedrock;

const enableGoogle = (s: GlobalStore) => modelProvider(s).google.enabled;
const googleAPIKey = (s: GlobalStore) => modelProvider(s).google.apiKey;
const googleProxyUrl = (s: GlobalStore) => modelProvider(s).google.endpoint;

const enableAzure = (s: GlobalStore) => modelProvider(s).openAI.useAzure;
const azureConfig = (s: GlobalStore) => modelProvider(s).azure;

const enableMistral = (s: GlobalStore) => modelProvider(s).mistral.enabled;
const mistralAPIKey = (s: GlobalStore) => modelProvider(s).mistral.apiKey;

const enableMoonshot = (s: GlobalStore) => modelProvider(s).moonshot.enabled;
const moonshotAPIKey = (s: GlobalStore) => modelProvider(s).moonshot.apiKey;

const enableOllamaConfigInSettings = (s: GlobalStore) =>
  s.serverConfig.languageModel?.ollama?.enabled || false;

const enableOllama = (s: GlobalStore) => modelProvider(s).ollama.enabled;
const ollamaProxyUrl = (s: GlobalStore) => modelProvider(s).ollama.endpoint;

const enablePerplexity = (s: GlobalStore) => modelProvider(s).perplexity.enabled;
const perplexityAPIKey = (s: GlobalStore) => modelProvider(s).perplexity.apiKey;

const enableAnthropic = (s: GlobalStore) => modelProvider(s).anthropic.enabled;
const anthropicAPIKey = (s: GlobalStore) => modelProvider(s).anthropic.apiKey;

// const azureModelList = (s: GlobalStore): ModelProviderCard => {
//   const azure = azureConfig(s);
//   return {
//     chatModels: parseModelString(azure.deployments),
//     id: 'azure',
//   };
// };

// 提取处理 chatModels 的专门方法
const processChatModels = (
  modelConfig: ReturnType<typeof parseModelString>,
  defaultChartModels = OpenAIProvider.chatModels,
): ChatModelCard[] => {
  let chatModels = modelConfig.removeAll ? [] : defaultChartModels;

  // 处理移除逻辑
  if (!modelConfig.removeAll) {
    chatModels = chatModels.filter((m) => !modelConfig.removed.includes(m.id));
  }

  return produce(chatModels, (draft) => {
    // 处理添加或替换逻辑
    for (const toAddModel of modelConfig.add) {
      // first try to find the model in LOBE_DEFAULT_MODEL_LIST to confirm if it is a known model
      const knownModel = LOBE_DEFAULT_MODEL_LIST.find((model) => model.id === toAddModel.id);

      // if the model is known, update it based on the known model
      if (knownModel) {
        const modelInList = draft.find((model) => model.id === toAddModel.id);

        // if the model is already in chatModels, update it
        if (modelInList) {
          if (modelInList.hidden) delete modelInList.hidden;
          if (toAddModel.displayName) modelInList.displayName = toAddModel.displayName;
        } else {
          // if the model is not in chatModels, add it
          draft.push({
            ...knownModel,
            displayName: toAddModel.displayName || knownModel.displayName || knownModel.id,
            hidden: undefined,
          });
        }
      } else {
        // if the model is not in LOBE_DEFAULT_MODEL_LIST, add it as a new custom model
        draft.push({
          ...toAddModel,
          displayName: toAddModel.displayName || toAddModel.id,
          functionCall: true,
          isCustom: true,
          vision: true,
        });
      }
    }
  });
};

const modelSelectList = (s: GlobalStore): ModelProviderCard[] => {
  const string = [
    s.serverConfig.customModelName,
    currentSettings(s).languageModel.openAI.customModelName,
  ]
    .filter(Boolean)
    .join(',');

  const modelConfig = parseModelString(string);

  const chatModels = processChatModels(modelConfig);

  const ollamaModelConfig = parseModelString(
    currentSettings(s).languageModel.ollama.customModelName,
  );

  const ollamaChatModels = processChatModels(ollamaModelConfig, OllamaProvider.chatModels);

  return [
    {
      ...OpenAIProvider,
      chatModels,
    },
    // { ...azureModelList(s), enabled: enableAzure(s) },
    { ...ZhiPuProvider, enabled: enableZhipu(s) },
    { ...MoonshotProvider, enabled: enableMoonshot(s) },
    { ...GoogleProvider, enabled: enableGoogle(s) },
    { ...BedrockProvider, enabled: enableBedrock(s) },
    { ...OllamaProvider, chatModels: ollamaChatModels, enabled: enableOllama(s) },
    { ...PerplexityProvider, enabled: enablePerplexity(s) },
    { ...AnthropicProvider, enabled: enableAnthropic(s) },
    { ...MistralProvider, enabled: enableMistral(s) },
  ];
};

const modelCardById = (id: string) => (s: GlobalStore) => {
  const list = modelSelectList(s);

  return list.flatMap((i) => i.chatModels).find((m) => m.id === id);
};

const modelEnabledFunctionCall = (id: string) => (s: GlobalStore) =>
  modelCardById(id)(s)?.functionCall || false;

// vision model white list, these models will change the content from string to array
// refs: https://github.com/lobehub/lobe-chat/issues/790
const modelEnabledVision = (id: string) => (s: GlobalStore) =>
  modelCardById(id)(s)?.vision || id.includes('vision');

const modelEnabledFiles = (id: string) => (s: GlobalStore) => modelCardById(id)(s)?.files;

const modelEnabledUpload = (id: string) => (s: GlobalStore) =>
  modelEnabledVision(id)(s) || modelEnabledFiles(id)(s);

const modelHasMaxToken = (id: string) => (s: GlobalStore) =>
  typeof modelCardById(id)(s)?.tokens !== 'undefined';

const modelMaxToken = (id: string) => (s: GlobalStore) => modelCardById(id)(s)?.tokens || 0;

/* eslint-disable sort-keys-fix/sort-keys-fix,  */
export const modelProviderSelectors = {
  modelSelectList,

  modelCardById,
  modelMaxToken,
  modelHasMaxToken,

  modelEnabledFunctionCall,
  modelEnabledVision,
  modelEnabledFiles,
  modelEnabledUpload,

  modelProviderConfig: modelProvider,
  providerEnabled,

  // OpenAI
  openAIConfig,
  openAIAPIKey,
  openAIProxyUrl,
  // Azure OpenAI
  enableAzure,
  azureConfig,
  // Zhipu
  enableZhipu,
  zhipuAPIKey,
  zhipuProxyUrl,
  // Google
  enableGoogle,
  googleAPIKey,
  googleProxyUrl,
  // Bedrock
  enableBedrock,
  bedrockConfig,

  // Moonshot
  enableMoonshot,
  moonshotAPIKey,

  // Ollama
  enableOllamaConfigInSettings,
  enableOllama,
  ollamaProxyUrl,

  // Perplexity
  enablePerplexity,
  perplexityAPIKey,

  // Anthropic
  enableAnthropic,
  anthropicAPIKey,
  
  // Mistral
  enableMistral,
  mistralAPIKey,
};
