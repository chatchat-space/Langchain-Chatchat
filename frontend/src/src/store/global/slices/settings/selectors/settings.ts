import { DEFAULT_LANG } from '@/const/locale';
import { DEFAULT_AGENT_META } from '@/const/meta';
import { DEFAULT_AGENT, DEFAULT_AGENT_CONFIG, DEFAULT_TTS_CONFIG } from '@/const/settings';
import { Locales } from '@/locales/resources';
import { GlobalSettings } from '@/types/settings';
import { isOnServerSide } from '@/utils/env';
import { merge } from '@/utils/merge';

import { GlobalStore } from '../../../store';

export const currentSettings = (s: GlobalStore): GlobalSettings =>
  merge(s.defaultSettings, s.settings);

const password = (s: GlobalStore) => currentSettings(s).password;

const currentTTS = (s: GlobalStore) => merge(DEFAULT_TTS_CONFIG, currentSettings(s).tts);

const defaultAgent = (s: GlobalStore) => merge(DEFAULT_AGENT, currentSettings(s).defaultAgent);

const defaultAgentConfig = (s: GlobalStore) => merge(DEFAULT_AGENT_CONFIG, defaultAgent(s).config);

const defaultAgentMeta = (s: GlobalStore) => merge(DEFAULT_AGENT_META, defaultAgent(s).meta);

// TODO: Maybe we can also export settings difference
const exportSettings = (s: GlobalStore) => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { password: _, ...settings } = currentSettings(s);

  return settings as GlobalSettings;
};

const currentLanguage = (s: GlobalStore) => {
  const locale = currentSettings(s).language;

  if (locale === 'auto') {
    if (isOnServerSide) return DEFAULT_LANG;

    return navigator.language as Locales;
  }

  return locale;
};

const dalleConfig = (s: GlobalStore) => currentSettings(s).tool?.dalle || {};
const isDalleAutoGenerating = (s: GlobalStore) => currentSettings(s).tool?.dalle?.autoGenerate;

export const settingsSelectors = {
  currentLanguage,
  currentSettings,
  currentTTS,
  dalleConfig,
  defaultAgent,
  defaultAgentConfig,
  defaultAgentMeta,
  exportSettings,
  isDalleAutoGenerating,
  password,
};
