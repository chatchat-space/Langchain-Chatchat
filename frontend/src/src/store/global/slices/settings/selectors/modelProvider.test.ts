import { describe, expect, it } from 'vitest';

import { DEFAULT_SETTINGS } from '@/const/settings';
import { agentSelectors } from '@/store/session/slices/agent';
import { merge } from '@/utils/merge';

import { GlobalStore, useGlobalStore } from '../../../store';
import { initialSettingsState } from '../initialState';
import { modelProviderSelectors } from './modelProvider';

describe('modelProviderSelectors', () => {
  describe('CUSTOM_MODELS', () => {
    it('custom deletion, addition, and renaming of models', () => {
      const s = merge(initialSettingsState, {
        serverConfig: {
          customModelName:
            '-all,+llama,+claude-2，-gpt-3.5-turbo,gpt-4-0125-preview=gpt-4-turbo,gpt-4-0125-preview=gpt-4-32k',
        },
      }) as unknown as GlobalStore;

      const result = modelProviderSelectors.modelSelectList(s).filter((r) => r.enabled);

      expect(result).toMatchSnapshot();
    });

    it('should work correct with gpt-4', () => {
      const s = merge(initialSettingsState, {
        serverConfig: {
          customModelName:
            '-all,+gpt-3.5-turbo-1106,+gpt-3.5-turbo,+gpt-3.5-turbo-16k,+gpt-4,+gpt-4-32k,+gpt-4-1106-preview,+gpt-4-vision-preview',
        },
      }) as unknown as GlobalStore;

      const result = modelProviderSelectors.modelSelectList(s).filter((r) => r.enabled);

      expect(result[0].chatModels).toMatchSnapshot();
    });
    it('duplicate naming model', () => {
      const s = merge(initialSettingsState, {
        serverConfig: {},
        settings: {
          languageModel: {
            openAI: {
              customModelName: 'gpt-4-0125-preview=gpt-4-turbo，gpt-4-0125-preview=gpt-4-32k',
            },
          },
        },
      }) as unknown as GlobalStore;

      const result = modelProviderSelectors.modelSelectList(s).filter((r) => r.enabled);

      expect(result[0].chatModels.find((s) => s.id === 'gpt-4-0125-preview')?.displayName).toEqual(
        'gpt-4-32k',
      );
    });

    it('should delete model', () => {
      const s = merge(initialSettingsState, {
        serverConfig: { customModelName: '-gpt-4' },
      }) as unknown as GlobalStore;

      const result = modelProviderSelectors.modelSelectList(s).filter((r) => r.enabled);

      expect(result.find((r) => r.id === 'gpt-4')).toBeUndefined();
    });

    it('show the hidden model', () => {
      const s = merge(initialSettingsState, {
        serverConfig: {},
        settings: {
          languageModel: {
            openAI: {
              customModelName: '+gpt-4-1106-preview',
            },
          },
        },
      }) as unknown as GlobalStore;

      const result = modelProviderSelectors.modelSelectList(s).filter((r) => r.enabled);

      expect(result[0].chatModels.find((o) => o.id === 'gpt-4-1106-preview')).toEqual({
        displayName: 'GPT-4 Turbo Preview (1106)',
        functionCall: true,
        id: 'gpt-4-1106-preview',
        tokens: 128000,
      });
    });

    it('only add the model', () => {
      const s = merge(initialSettingsState, {
        serverConfig: {},
        settings: {
          languageModel: {
            openAI: {
              customModelName: 'model1,model2,model3，model4',
            },
          },
        },
      }) as unknown as GlobalStore;

      const result = modelProviderSelectors.modelSelectList(s).filter((r) => r.enabled);

      expect(result[0].chatModels).toContainEqual({
        displayName: 'model1',
        functionCall: true,
        id: 'model1',
        isCustom: true,
        vision: true,
      });
      expect(result[0].chatModels).toContainEqual({
        displayName: 'model2',
        functionCall: true,
        id: 'model2',
        isCustom: true,
        vision: true,
      });
      expect(result[0].chatModels).toContainEqual({
        displayName: 'model3',
        functionCall: true,
        id: 'model3',
        isCustom: true,
        vision: true,
      });
      expect(result[0].chatModels).toContainEqual({
        displayName: 'model4',
        functionCall: true,
        id: 'model4',
        isCustom: true,
        vision: true,
      });
    });
  });

  describe('modelEnabledVision', () => {
    it('should return true if the model has vision ability', () => {
      const hasAbility = modelProviderSelectors.modelEnabledVision('gpt-4-vision-preview')(
        useGlobalStore.getState(),
      );
      expect(hasAbility).toBeTruthy();
    });

    it('should return false if the model does not have vision ability', () => {
      const hasAbility = modelProviderSelectors.modelEnabledVision('some-other-model')(
        useGlobalStore.getState(),
      );

      expect(hasAbility).toBeFalsy();
    });

    it('should return false if the model include vision in id', () => {
      const hasAbility = modelProviderSelectors.modelEnabledVision('some-other-model-vision')(
        useGlobalStore.getState(),
      );

      expect(hasAbility).toBeTruthy();
    });
  });

  describe('modelEnabledFiles', () => {
    it('should return false if the model does not have file ability', () => {
      const enabledFiles = modelProviderSelectors.modelEnabledFiles('gpt-4-vision-preview')(
        useGlobalStore.getState(),
      );
      expect(enabledFiles).toBeFalsy();
    });

    it('should return true if the model has file ability', () => {
      const enabledFiles = modelProviderSelectors.modelEnabledFiles('gpt-4-all')(
        useGlobalStore.getState(),
      );
      expect(enabledFiles).toBeTruthy();
    });
  });

  describe('modelHasMaxToken', () => {
    it('should return true if the model is in the list of models that show tokens', () => {
      const show = modelProviderSelectors.modelHasMaxToken('gpt-3.5-turbo')(
        useGlobalStore.getState(),
      );
      expect(show).toBeTruthy();
    });

    it('should return false if the model is not in the list of models that show tokens', () => {
      const show = modelProviderSelectors.modelHasMaxToken('some-other-model')(
        useGlobalStore.getState(),
      );
      expect(show).toBe(false);
    });
  });
});
