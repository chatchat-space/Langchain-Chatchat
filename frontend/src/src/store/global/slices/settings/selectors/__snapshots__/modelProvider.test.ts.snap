// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`modelProviderSelectors > CUSTOM_MODELS > custom deletion, addition, and renaming of models 1`] = `
[
  {
    "chatModels": [
      {
        "displayName": "llama",
        "functionCall": true,
        "id": "llama",
        "isCustom": true,
        "vision": true,
      },
      {
        "displayName": "claude-2",
        "functionCall": true,
        "id": "claude-2",
        "isCustom": true,
        "vision": true,
      },
      {
        "displayName": "gpt-4-32k",
        "functionCall": true,
        "hidden": undefined,
        "id": "gpt-4-0125-preview",
        "tokens": 128000,
      },
    ],
    "enabled": true,
    "id": "openai",
  },
]
`;

exports[`modelProviderSelectors > CUSTOM_MODELS > should work correct with gpt-4 1`] = `
[
  {
    "displayName": "GPT-3.5 Turbo (1106)",
    "functionCall": true,
    "hidden": undefined,
    "id": "gpt-3.5-turbo-1106",
    "tokens": 16385,
  },
  {
    "description": "GPT 3.5 Turbo，适用于各种文本生成和理解任务",
    "displayName": "GPT-3.5 Turbo",
    "functionCall": true,
    "hidden": undefined,
    "id": "gpt-3.5-turbo",
    "tokens": 16385,
  },
  {
    "displayName": "GPT-3.5 Turbo 16K",
    "hidden": undefined,
    "id": "gpt-3.5-turbo-16k",
    "tokens": 16385,
  },
  {
    "displayName": "GPT-4",
    "functionCall": true,
    "hidden": undefined,
    "id": "gpt-4",
    "tokens": 8192,
  },
  {
    "displayName": "gpt-4-32k",
    "functionCall": true,
    "hidden": undefined,
    "id": "gpt-4-32k",
    "tokens": 32768,
  },
  {
    "displayName": "GPT-4 Turbo Preview (1106)",
    "functionCall": true,
    "hidden": undefined,
    "id": "gpt-4-1106-preview",
    "tokens": 128000,
  },
  {
    "description": "GPT-4 视觉预览版，支持视觉任务",
    "displayName": "GPT-4 Turbo Vision Preview",
    "hidden": undefined,
    "id": "gpt-4-vision-preview",
    "tokens": 128000,
    "vision": true,
  },
]
`;
